src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java

```diff
@@ -0,0 +1,374 @@
+
+        switch (columnsExpression.kind())
+        {
+            case SINGLE_COLUMN:
+                List<ByteBuffer> buffers = bindAndGet(options);
+
+                ColumnMetadata column = firstColumn();
+                if (operator == Operator.IN)
+                {
+                    filter.add(column, operator, inValues(column, buffers));
+                }
+                else if (operator == Operator.LIKE)
+                {
+                    LikePattern pattern = LikePattern.parse(buffers.get(0));
+                    // there must be a suitable INDEX for LIKE_XXX expressions
+                    RowFilter.SimpleExpression expression = filter.add(column, pattern.kind().operator(), pattern.value());
+                    indexRegistry.getBestIndexFor(expression)
+                                 .orElseThrow(() -> invalidRequest("%s is only supported on properly indexed columns",
+                                                                   expression));
+                }
+                else
+                {
+                    filter.add(column, operator, buffers.get(0));
+                }
+                break;
+            case MULTI_COLUMN:
+                checkFalse(isSlice(), "Multi-column slice restrictions cannot be used for filtering.");
+
+                if (isEQ())
+                {
+                    List<ByteBuffer> elements = bindAndGetElements(options).get(0);
+
+                    for (int i = 0, m = columns().size(); i < m; i++)
+                    {
+                        ColumnMetadata columnDef = columns().get(i);
+                        filter.add(columnDef, Operator.EQ, elements.get(i));
+                    }
+                }
+                else if (isIN())
+                {
+                    // If the relation is of the type (c) IN ((x),(y),(z)) then it is equivalent to
+                    // c IN (x, y, z) and we can perform filtering
+                    if (columns().size() == 1)
+                    {
+                        List<ByteBuffer> values = bindAndGetElements(options).stream()
+                                                                             .map(elements -> elements.get(0))
+                                                                             .collect(Collectors.toList());
+
+                        filter.add(firstColumn(), Operator.IN, inValues(firstColumn(), values));
+                    }
+                    else
+                    {
+                        throw invalidRequest("Multicolumn IN filters are not supported");
+                    }
+                }
+                break;
+            case MAP_ELEMENT:
+                ByteBuffer key = columnsExpression.mapKey(options);
+                List<ByteBuffer> values = bindAndGet(options);
+                filter.addMapEquality(firstColumn(), key, operator, values.get(0));
+                break;
+            default: throw new UnsupportedOperationException();
```

Comment:
You can ignore everything that is related to IS NOT NULL because that code was never used. Nothing will prevent an index to support those operators once they will be properly supported by C*.